docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,616 @@
+
+### 4.1.6 Delete-Delete Contention
+
+There are two cases for delete-delete contention: either the two time ranges contend over
+the same pointer(s), or they are disjoint: the case where they are disjoint is simple: we
+can simply use Mutex Locks to turn the operations into serial operations to the index and
+there would be no conflicts (see the detailed deletion section 4.2.3).
+
+If the two deletions contend over the same pointers, there are two more cases: the two
+time ranges could overlap each other, in which case the controller at the Unary level would
+return an error and disallow the deletion, or they don't overlap each other, in which case
+the operation is turned into a Serial operation of the index (see section 4.2.3).
+
+The following diagram summarizes possible conflicts between these entities:
+
+<div align="center">
+    <img src="img/0019-240529-cesium-race-conditions/Entity-Interactions.png">
+    <h6>Cesium entities' interaction</h6>
+</div>
+
+## 4.2 Domain level race conditions
+
+We will now inspect, from the bottom up, how the various entities interact with the
+domain data and the index in memory. The domain level is responsible for managing the
+underlying file system and the index. A domainDB is index-agnostic: it knows nothing
+about the time stamps that these samples correspond to.
+
+Dimensional analysis is a great tool to reveal nature of data we are dealing with: it
+is very helpful to remember that data written and read from at the domain level are simple
+**byte arrays**.
+
+### 4.2.1 Writing
+
+A write proceeds in three phases:
+1. Acquire a writer (getting a file handle to an existing file or opening a new one)
+to an underlying file that still has space for more data.
+2. Write the binary telemetry data to the file handle.
+3. Commit the writer, i.e. add a pointer in the index storing the time range represented
+by the data and the location of the data in the file system (file key, offset, length).
+
+Cesium is different from other Database systems in that a writer may be
+committed multiple times: a writer may repeat steps 2 and 3, i.e.
+
+4. Write more binary telemetry data to the file handle
+5. Commit the writer, i.e. update the pointer that previously described the domain
+written by this writer to contain the new domain.
+
+In Synnax version 19, file cutoffs were introduced, which upon a writer writing to a file
+that reaches its size limit, automatically ends the domain and begins a new one. We must
+also consider the creation of this new pointer when analyzing for race conditions.
+
+#### 4.2.1.1 Acquiring a writer
+
+When acquiring a writer, we must guarantee that we acquire a file handle on a file that
+is both exclusive and under the file size limit. We first scan through the open file handles
+in the file controller under a `RLock` and upon finding a file within the size limit, we
+try to acquire it by comparing-and-swapping its flag (`True` indicates the writer is currently
```

Comment:
I don't really get the point of saying things like `True` indicates the specific flag state if you don't either a) give an overview of the structure of a writer or b) give code above to show what the writer looks like. I would do a) instead of b).